<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />

<meta name="author" content="Luca Lepore" />

<meta name="date" content="2024-08-12" />

<title>Getting Started with GenePack</title>

<script>// Pandoc 2.9 adds attributes on both header and div. We remove the former (to
// be compatible with the behavior of Pandoc < 2.8).
document.addEventListener('DOMContentLoaded', function(e) {
  var hs = document.querySelectorAll("div.section[class*='level'] > :first-child");
  var i, h, a;
  for (i = 0; i < hs.length; i++) {
    h = hs[i];
    if (!/^h[1-6]$/i.test(h.tagName)) continue;  // it should be a header h1-h6
    a = h.attributes;
    while (a.length > 0) h.removeAttribute(a[0].name);
  }
});
</script>

<style type="text/css">
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
</style>



<style type="text/css">
code {
white-space: pre;
}
.sourceCode {
overflow: visible;
}
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
{ counter-reset: source-line 0; }
pre.numberSource code > span
{ position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
{ content: counter(source-line);
position: relative; left: -1em; text-align: right; vertical-align: baseline;
border: none; display: inline-block;
-webkit-touch-callout: none; -webkit-user-select: none;
-khtml-user-select: none; -moz-user-select: none;
-ms-user-select: none; user-select: none;
padding: 0 4px; width: 4em;
color: #aaaaaa;
}
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa; padding-left: 4px; }
div.sourceCode
{ }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } 
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.at { color: #7d9029; } 
code span.bn { color: #40a070; } 
code span.bu { color: #008000; } 
code span.cf { color: #007020; font-weight: bold; } 
code span.ch { color: #4070a0; } 
code span.cn { color: #880000; } 
code span.co { color: #60a0b0; font-style: italic; } 
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.do { color: #ba2121; font-style: italic; } 
code span.dt { color: #902000; } 
code span.dv { color: #40a070; } 
code span.er { color: #ff0000; font-weight: bold; } 
code span.ex { } 
code span.fl { color: #40a070; } 
code span.fu { color: #06287e; } 
code span.im { color: #008000; font-weight: bold; } 
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.kw { color: #007020; font-weight: bold; } 
code span.op { color: #666666; } 
code span.ot { color: #007020; } 
code span.pp { color: #bc7a00; } 
code span.sc { color: #4070a0; } 
code span.ss { color: #bb6688; } 
code span.st { color: #4070a0; } 
code span.va { color: #19177c; } 
code span.vs { color: #4070a0; } 
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } 
</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    var j = 0;
    while (j < rules.length) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") {
        j++;
        continue;
      }
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') {
        j++;
        continue;
      }
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">Getting Started with GenePack</h1>
<h4 class="author">Luca Lepore</h4>
<h4 class="date">2024-08-12</h4>



<div id="introduction-to-genepack" class="section level1">
<h1>Introduction to GenePack</h1>
<p>The <strong>GenePack</strong> package provides a comprehensive set of
S4 classes to represent various human gene types, including both
protein-coding and non-coding genes. The package includes constructors,
validators, and accessor functions to manage and modify the gene objects
created. Each gene object can store its essential information,
particularly the Ensembl ID, HUGO symbol, name, description, gene
structure, and gene product (the product ID and sequence). In this
implementation, each transcript isoform is treated as a separate gene,
using the Ensembl transcript ID instead of the Ensembl gene ID to
simplify the representation and management of gene products, and provide
more flexibility.</p>
<p>The classes provided by GenePack are organized as follows:
‘ProteinCodingGene’; ‘LncRNAGene’; ‘HousekeepingRNAGene’: ‘RRNAGene’,
‘TRNAGene’; ‘SncRNAGene’: ‘MicroRNAGene’, ‘SiRNAGene’, ‘PiRNAGene’,
‘SnRNAGene’, ‘SnoRNAGene’.</p>
<p>The HousekeepingRNAGene and SncRNAGene classes are virtual classes
that inherit from the Gene class. They are provided to biologically
group their derived classes. This design makes easier potential future
updates and the introduction of new functionalities.</p>
<p>This vignette will guide the user in the usage of the GenePack
package, showing how to create and manipulate gene objects, validate
gene products, and manage various gene types.</p>
</div>
<div id="installation" class="section level1">
<h1>Installation</h1>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="fu">library</span>(GenePack)</span></code></pre></div>
</div>
<div id="creating-a-gene-object" class="section level1">
<h1>Creating a Gene object</h1>
<p>All the gene classes available in this package are specialized
classes deriving from a general virtual Gene class. The Gene class sets
the common parameters among all the derived classes with the related
internal validations. These include the Ensembl transcript ID, HUGO
Symbol, name, description, tissue specificity, gene structure, and
clinical significance. The gene product argument is class-specific and
managed within each derived class instead.</p>
<p>The ‘gene_structure’ and ‘gene_product’ arguments are complex object
and should be created separately by the user before constructing the
Gene object. This approach maximizes flexibility, allowing the user to
better exploit the functionalities provided by the GenomicRanges package
without predefined constraints in the constructor. By separating the
creation of these parameters, the creation of the gene object becomes
more manageable and improves mantainability of the code. In addition,
the possibility to reuse the ‘gene_structure’ and ‘gene_product’ objects
in other contexts avoids code duplications and increases efficiency.</p>
<div id="creating-the-granges-object" class="section level2">
<h2>Creating the GRanges object</h2>
<p>An example regarding the creation of a <strong>GRanges</strong>
object (the gene structure) is shown below. The
<strong>GenomicRanges</strong> library is included in the package and
can be directly used as follows or it can be explicitly imported calling
library(GenomicRanges). In this case, the GRanges object belongs to a
long non-coding rna. After the creation, it can be passed in the
constructor of the Gene class to create the Gene object.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a>lncrnagene_structure <span class="ot">&lt;-</span> GenomicRanges<span class="sc">::</span><span class="fu">GRanges</span>(<span class="at">seqnames =</span> <span class="st">&quot;chr1&quot;</span>,</span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a>                                                 <span class="at">ranges =</span> IRanges<span class="sc">::</span><span class="fu">IRanges</span>(<span class="at">start =</span> <span class="dv">200</span>, <span class="at">end =</span> <span class="dv">1200</span>), </span>
<span id="cb2-3"><a href="#cb2-3" tabindex="-1"></a>                                                 <span class="at">strand =</span> <span class="st">&quot;+&quot;</span>)</span></code></pre></div>
</div>
<div id="creating-the-gene_product-list-and-the-final-gene-object" class="section level2">
<h2>Creating the gene_product list and the final Gene object</h2>
<p>An example regarding the creation of the gene product and the final
Gene object is shown below. All the parameters are provided, but just
‘id’, ‘hugo_symbol’, ‘gene_structure’, ‘gene_product’ are mandatory.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a>lncrnagene_product <span class="ot">&lt;-</span> <span class="fu">list</span>(<span class="at">lncrna_id =</span> <span class="st">&quot;lncRNAID&quot;</span>, </span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a>                             <span class="at">lncrna_sequence =</span> </span>
<span id="cb3-3"><a href="#cb3-3" tabindex="-1"></a>                                <span class="fu">paste0</span>(<span class="st">&quot;AUGCUUAGCGUACGGUAGGCUUAACUGCGUACGAUCGAUCGGCUAAUCGGCUUAGCGUA&quot;</span>,</span>
<span id="cb3-4"><a href="#cb3-4" tabindex="-1"></a>                                       <span class="st">&quot;CGGUAGGCUUAACUGCGUACGAUCGAUCGGCUAAGCGUACGGUAGGCUUAACUGCGUAC&quot;</span>,</span>
<span id="cb3-5"><a href="#cb3-5" tabindex="-1"></a>                                       <span class="st">&quot;GAUCGAUCGGCUUAGCGUACGUAGGCUCAACUGCGUACGAUCGAUCGGCUUAGCGUACG&quot;</span>,</span>
<span id="cb3-6"><a href="#cb3-6" tabindex="-1"></a>                                       <span class="st">&quot;GUAGGCUUAACUGCGGACGAUCGAUCGGCUUAGCGUACGGUAGGCUUAACUGCGUACGA&quot;</span>,</span>
<span id="cb3-7"><a href="#cb3-7" tabindex="-1"></a>                                       <span class="st">&quot;UCGAUCGC&quot;</span>))</span>
<span id="cb3-8"><a href="#cb3-8" tabindex="-1"></a>  </span>
<span id="cb3-9"><a href="#cb3-9" tabindex="-1"></a>lncrna_gene <span class="ot">&lt;-</span> GenePack<span class="sc">::</span><span class="fu">createLncRNAGene</span>(<span class="at">id =</span> <span class="st">&quot;ENST00005859745&quot;</span>,</span>
<span id="cb3-10"><a href="#cb3-10" tabindex="-1"></a>                                          <span class="at">hugo_symbol =</span> <span class="st">&quot;SYMBOL1&quot;</span>,</span>
<span id="cb3-11"><a href="#cb3-11" tabindex="-1"></a>                                          <span class="at">name =</span> <span class="st">&quot;lncRNA gene name&quot;</span>,</span>
<span id="cb3-12"><a href="#cb3-12" tabindex="-1"></a>                                          <span class="at">description =</span> <span class="st">&quot;gene description&quot;</span>,</span>
<span id="cb3-13"><a href="#cb3-13" tabindex="-1"></a>                                          <span class="at">tissue_specificity =</span> <span class="fu">list</span>(<span class="st">&quot;liver&quot;</span>, <span class="st">&quot;small bowel&quot;</span>),</span>
<span id="cb3-14"><a href="#cb3-14" tabindex="-1"></a>                                          <span class="at">gene_structure =</span> lncrnagene_structure,</span>
<span id="cb3-15"><a href="#cb3-15" tabindex="-1"></a>                                          <span class="at">gene_product =</span> lncrnagene_product,</span>
<span id="cb3-16"><a href="#cb3-16" tabindex="-1"></a>                                          <span class="at">clinical_significance =</span> <span class="st">&quot;association with bowel cancer&quot;</span>)</span></code></pre></div>
</div>
<div id="further-examples" class="section level2">
<h2>Further examples</h2>
<p>One example for each main Gene class is provided below.</p>
<p><strong>ProteinCodingGene object</strong></p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a>codinggene_structure <span class="ot">&lt;-</span> GenomicRanges<span class="sc">::</span><span class="fu">GRanges</span>(<span class="at">seqnames =</span> <span class="st">&quot;chr1&quot;</span>,</span>
<span id="cb4-2"><a href="#cb4-2" tabindex="-1"></a>                                                 <span class="at">ranges =</span> IRanges<span class="sc">::</span><span class="fu">IRanges</span>(<span class="at">start =</span> <span class="dv">200</span>, <span class="at">end =</span> <span class="dv">1200</span>), </span>
<span id="cb4-3"><a href="#cb4-3" tabindex="-1"></a>                                                 <span class="at">strand =</span> <span class="st">&quot;+&quot;</span>)</span>
<span id="cb4-4"><a href="#cb4-4" tabindex="-1"></a>  </span>
<span id="cb4-5"><a href="#cb4-5" tabindex="-1"></a>codinggene_product <span class="ot">&lt;-</span> <span class="fu">list</span>(<span class="at">protein_id =</span> <span class="st">&quot;proteinID&quot;</span>, </span>
<span id="cb4-6"><a href="#cb4-6" tabindex="-1"></a>                           <span class="at">protein_sequence =</span> </span>
<span id="cb4-7"><a href="#cb4-7" tabindex="-1"></a>                             <span class="fu">paste0</span>(<span class="st">&quot;MEEPQSDPSVEPPLSQETFSDLWKLLPENNVTELTKVEPPEVPVNGPAAANLL&quot;</span>,</span>
<span id="cb4-8"><a href="#cb4-8" tabindex="-1"></a>                                    <span class="st">&quot;GGQMKKLGFLHSGTAKSVTCTYSPALNKLTENSSEKGSITRRRCFEGIKTM&quot;</span>))</span>
<span id="cb4-9"><a href="#cb4-9" tabindex="-1"></a>  </span>
<span id="cb4-10"><a href="#cb4-10" tabindex="-1"></a>proteincoding_gene <span class="ot">&lt;-</span> GenePack<span class="sc">::</span><span class="fu">createProteinCodingGene</span>(<span class="at">id =</span> <span class="st">&quot;ENST00005859745&quot;</span>,</span>
<span id="cb4-11"><a href="#cb4-11" tabindex="-1"></a>                                            <span class="at">hugo_symbol =</span> <span class="st">&quot;SYMBOL1&quot;</span>,</span>
<span id="cb4-12"><a href="#cb4-12" tabindex="-1"></a>                                            <span class="at">name =</span> <span class="st">&quot;protein-coding gene name&quot;</span>,</span>
<span id="cb4-13"><a href="#cb4-13" tabindex="-1"></a>                                            <span class="at">description =</span> <span class="st">&quot;gene description&quot;</span>,</span>
<span id="cb4-14"><a href="#cb4-14" tabindex="-1"></a>                                            <span class="at">tissue_specificity =</span> <span class="fu">list</span>(<span class="st">&quot;liver&quot;</span>, <span class="st">&quot;small bowel&quot;</span>),</span>
<span id="cb4-15"><a href="#cb4-15" tabindex="-1"></a>                                            <span class="at">gene_structure =</span> codinggene_structure,</span>
<span id="cb4-16"><a href="#cb4-16" tabindex="-1"></a>                                            <span class="at">gene_product =</span> codinggene_product,</span>
<span id="cb4-17"><a href="#cb4-17" tabindex="-1"></a>                                            <span class="at">clinical_significance =</span> <span class="st">&quot;association with bowel cancer&quot;</span>)</span></code></pre></div>
<p><strong>MicroRNAGene object</strong></p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a>micrornagene_structure <span class="ot">&lt;-</span> GenomicRanges<span class="sc">::</span><span class="fu">GRanges</span>(<span class="at">seqnames =</span> <span class="st">&quot;chr1&quot;</span>,</span>
<span id="cb5-2"><a href="#cb5-2" tabindex="-1"></a>                                                   <span class="at">ranges =</span> IRanges<span class="sc">::</span><span class="fu">IRanges</span>(<span class="at">start =</span> <span class="dv">200</span>, <span class="at">end =</span> <span class="dv">1200</span>), </span>
<span id="cb5-3"><a href="#cb5-3" tabindex="-1"></a>                                                   <span class="at">strand =</span> <span class="st">&quot;+&quot;</span>)</span>
<span id="cb5-4"><a href="#cb5-4" tabindex="-1"></a>  </span>
<span id="cb5-5"><a href="#cb5-5" tabindex="-1"></a>micrornagene_product <span class="ot">&lt;-</span> <span class="fu">list</span>(<span class="at">microrna_id =</span> <span class="st">&quot;micrornaID&quot;</span>, </span>
<span id="cb5-6"><a href="#cb5-6" tabindex="-1"></a>                             <span class="at">microrna_sequence =</span> <span class="st">&quot;UGAGGUAGUAGGUUGUAUAGUU&quot;</span>)</span>
<span id="cb5-7"><a href="#cb5-7" tabindex="-1"></a>  </span>
<span id="cb5-8"><a href="#cb5-8" tabindex="-1"></a>microrna_gene <span class="ot">&lt;-</span> GenePack<span class="sc">::</span><span class="fu">createMicroRNAGene</span>(<span class="at">id =</span> <span class="st">&quot;ENST00005859745&quot;</span>,</span>
<span id="cb5-9"><a href="#cb5-9" tabindex="-1"></a>                                                <span class="at">hugo_symbol =</span> <span class="st">&quot;SYMBOL1&quot;</span>,</span>
<span id="cb5-10"><a href="#cb5-10" tabindex="-1"></a>                                                <span class="at">name =</span> <span class="st">&quot;microRNA gene name&quot;</span>,</span>
<span id="cb5-11"><a href="#cb5-11" tabindex="-1"></a>                                                <span class="at">description =</span> <span class="st">&quot;gene description&quot;</span>,</span>
<span id="cb5-12"><a href="#cb5-12" tabindex="-1"></a>                                                <span class="at">tissue_specificity =</span> <span class="fu">list</span>(<span class="st">&quot;liver&quot;</span>, <span class="st">&quot;small bowel&quot;</span>),</span>
<span id="cb5-13"><a href="#cb5-13" tabindex="-1"></a>                                                <span class="at">gene_structure =</span> micrornagene_structure,</span>
<span id="cb5-14"><a href="#cb5-14" tabindex="-1"></a>                                                <span class="at">gene_product =</span> micrornagene_product,</span>
<span id="cb5-15"><a href="#cb5-15" tabindex="-1"></a>                                                <span class="at">clinical_significance =</span> <span class="st">&quot;association with bowel cancer&quot;</span>)</span></code></pre></div>
<p><strong>RRNAGene object</strong></p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a>rrnagene_structure <span class="ot">&lt;-</span> GenomicRanges<span class="sc">::</span><span class="fu">GRanges</span>(<span class="at">seqnames =</span> <span class="st">&quot;chr1&quot;</span>,</span>
<span id="cb6-2"><a href="#cb6-2" tabindex="-1"></a>                                               <span class="at">ranges =</span> IRanges<span class="sc">::</span><span class="fu">IRanges</span>(<span class="at">start =</span> <span class="dv">200</span>, <span class="at">end =</span> <span class="dv">1200</span>), </span>
<span id="cb6-3"><a href="#cb6-3" tabindex="-1"></a>                                               <span class="at">strand =</span> <span class="st">&quot;+&quot;</span>)</span>
<span id="cb6-4"><a href="#cb6-4" tabindex="-1"></a>  </span>
<span id="cb6-5"><a href="#cb6-5" tabindex="-1"></a>rrnagene_product <span class="ot">&lt;-</span> <span class="fu">list</span>(<span class="at">rrna_id =</span> <span class="st">&quot;rrnaID&quot;</span>, </span>
<span id="cb6-6"><a href="#cb6-6" tabindex="-1"></a>                         <span class="at">rrna_sequence =</span> </span>
<span id="cb6-7"><a href="#cb6-7" tabindex="-1"></a>                           <span class="fu">paste0</span>(<span class="st">&quot;GAGCCCGGCGUCCCGUCUCAGACCCGGCCGACAGGAGGGGUGGAGUGGGUGUGUGCGUG&quot;</span>,</span>
<span id="cb6-8"><a href="#cb6-8" tabindex="-1"></a>                                  <span class="st">&quot;UGGAGGAGGGUGUGGUGGUGCGGUGUGGUGGGCGUGUGAGUGUGUGGCGCGUGGAGGAG&quot;</span>))</span>
<span id="cb6-9"><a href="#cb6-9" tabindex="-1"></a>  </span>
<span id="cb6-10"><a href="#cb6-10" tabindex="-1"></a>rrna_gene <span class="ot">&lt;-</span> GenePack<span class="sc">::</span><span class="fu">createRRNAGene</span>(<span class="at">id =</span> <span class="st">&quot;ENST00005859745&quot;</span>,</span>
<span id="cb6-11"><a href="#cb6-11" tabindex="-1"></a>                                                    <span class="at">hugo_symbol =</span> <span class="st">&quot;SYMBOL1&quot;</span>,</span>
<span id="cb6-12"><a href="#cb6-12" tabindex="-1"></a>                                                    <span class="at">name =</span> <span class="st">&quot;rRNA gene name&quot;</span>,</span>
<span id="cb6-13"><a href="#cb6-13" tabindex="-1"></a>                                                    <span class="at">description =</span> <span class="st">&quot;gene description&quot;</span>,</span>
<span id="cb6-14"><a href="#cb6-14" tabindex="-1"></a>                                                    <span class="at">tissue_specificity =</span> <span class="fu">list</span>(<span class="st">&quot;liver&quot;</span>, <span class="st">&quot;small bowel&quot;</span>),</span>
<span id="cb6-15"><a href="#cb6-15" tabindex="-1"></a>                                                    <span class="at">gene_structure =</span> rrnagene_structure,</span>
<span id="cb6-16"><a href="#cb6-16" tabindex="-1"></a>                                                    <span class="at">gene_product =</span> rrnagene_product,</span>
<span id="cb6-17"><a href="#cb6-17" tabindex="-1"></a>                                                    <span class="at">clinical_significance =</span> <span class="st">&quot;association bowel cancer&quot;</span>)</span></code></pre></div>
</div>
</div>
<div id="accessing-and-modifying-gene-objects-information" class="section level1">
<h1>Accessing and modifying gene objects information</h1>
<p>GenePack also provides accessor functions to access and modify
information within the gene objects created in order to avoid directly
interactions with the objects’ slots. This guarantees a robust way to
update gene information preserving data integrity.</p>
<div id="examples" class="section level2">
<h2>Examples</h2>
<p>Examples regarding the mandatory arguments are provided below.</p>
<p><strong>Get and Set Ensembl ID</strong></p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a>GenePack<span class="sc">::</span><span class="fu">getID</span>(lncrna_gene)</span></code></pre></div>
<pre><code>## [1] &quot;ENST00005859745&quot;</code></pre>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" tabindex="-1"></a>GenePack<span class="sc">::</span><span class="fu">setID</span>(lncrna_gene) <span class="ot">&lt;-</span> <span class="st">&quot;ENST00005859750&quot;</span></span>
<span id="cb9-2"><a href="#cb9-2" tabindex="-1"></a>GenePack<span class="sc">::</span><span class="fu">getID</span>(lncrna_gene)</span></code></pre></div>
<pre><code>## [1] &quot;ENST00005859750&quot;</code></pre>
<p><strong>Get and Set HUGO Symbol</strong></p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" tabindex="-1"></a>GenePack<span class="sc">::</span><span class="fu">getHugoSymbol</span>(lncrna_gene)</span></code></pre></div>
<pre><code>## [1] &quot;SYMBOL1&quot;</code></pre>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" tabindex="-1"></a>GenePack<span class="sc">::</span><span class="fu">setHugoSymbol</span>(lncrna_gene) <span class="ot">&lt;-</span> <span class="st">&quot;NEWSYMBOL1&quot;</span></span>
<span id="cb13-2"><a href="#cb13-2" tabindex="-1"></a>GenePack<span class="sc">::</span><span class="fu">getHugoSymbol</span>(lncrna_gene)</span></code></pre></div>
<pre><code>## [1] &quot;NEWSYMBOL1&quot;</code></pre>
<p><strong>Get and Set Gene structure</strong></p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" tabindex="-1"></a>GenePack<span class="sc">::</span><span class="fu">getGeneStructure</span>(lncrna_gene)</span></code></pre></div>
<pre><code>## GRanges object with 1 range and 0 metadata columns:
##       seqnames    ranges strand
##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;
##   [1]     chr1  200-1200      +
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" tabindex="-1"></a>new_gene_structure2 <span class="ot">&lt;-</span> GenomicRanges<span class="sc">::</span><span class="fu">GRanges</span>(<span class="at">seqnames =</span> <span class="st">&quot;chr1&quot;</span>,</span>
<span id="cb17-2"><a href="#cb17-2" tabindex="-1"></a>                                               <span class="at">ranges =</span> IRanges<span class="sc">::</span><span class="fu">IRanges</span>(<span class="at">start =</span> <span class="dv">205</span>, <span class="at">end =</span> <span class="dv">1210</span>),</span>
<span id="cb17-3"><a href="#cb17-3" tabindex="-1"></a>                                               <span class="at">strand =</span> <span class="st">&quot;+&quot;</span>)</span>
<span id="cb17-4"><a href="#cb17-4" tabindex="-1"></a></span>
<span id="cb17-5"><a href="#cb17-5" tabindex="-1"></a>GenePack<span class="sc">::</span><span class="fu">setGeneStructure</span>(lncrna_gene) <span class="ot">&lt;-</span> new_gene_structure2</span>
<span id="cb17-6"><a href="#cb17-6" tabindex="-1"></a>GenePack<span class="sc">::</span><span class="fu">getGeneStructure</span>(lncrna_gene)</span></code></pre></div>
<pre><code>## GRanges object with 1 range and 0 metadata columns:
##       seqnames    ranges strand
##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;
##   [1]     chr1  205-1210      +
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
<p><strong>Get and Set product ID</strong></p>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" tabindex="-1"></a>GenePack<span class="sc">::</span><span class="fu">getProductID</span>(lncrna_gene)</span></code></pre></div>
<pre><code>## [1] &quot;lncRNAID&quot;</code></pre>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" tabindex="-1"></a>GenePack<span class="sc">::</span><span class="fu">setProductID</span>(lncrna_gene) <span class="ot">&lt;-</span> <span class="st">&quot;newlncRNAID&quot;</span></span>
<span id="cb21-2"><a href="#cb21-2" tabindex="-1"></a>GenePack<span class="sc">::</span><span class="fu">getProductID</span>(lncrna_gene)</span></code></pre></div>
<pre><code>## [1] &quot;newlncRNAID&quot;</code></pre>
<p><strong>Get and Set product sequence</strong></p>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" tabindex="-1"></a>GenePack<span class="sc">::</span><span class="fu">getProductSequence</span>(lncrna_gene)</span></code></pre></div>
<pre><code>## AUGCUUAGCGUACGGUAGGCUUAACUGCGUACGAUCGAUCGGCUAAUCGGCUUAGCGUACGGUAGGCUUAACUGCGUACG 
## AUCGAUCGGCUAAGCGUACGGUAGGCUUAACUGCGUACGAUCGAUCGGCUUAGCGUACGUAGGCUCAACUGCGUACGAUC 
## GAUCGGCUUAGCGUACGGUAGGCUUAACUGCGGACGAUCGAUCGGCUUAGCGUACGGUAGGCUUAACUGCGUACGAUCGA 
## UCGC</code></pre>
<div class="sourceCode" id="cb25"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" tabindex="-1"></a>seq <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">&quot;AUGCUUAGCGUACGGUAGGCUUAACUGCGUACGAUCGAUCGGCUAAUCGGCUUAGCGUACGGUAGG&quot;</span>,</span>
<span id="cb25-2"><a href="#cb25-2" tabindex="-1"></a>                <span class="st">&quot;CUUAACUGCGUACGAUCGAUCGGCUAAGCGUACGGUAGGCUUAACUGCGUACGAUCGAUCGGCUUAGCGUAC&quot;</span>,</span>
<span id="cb25-3"><a href="#cb25-3" tabindex="-1"></a>                <span class="st">&quot;GUAGGCUCAACUGCGUACGAUCGAUCGGCUUAGCGUACGGUAGGCUUAACUGCGGACGAUCGAUCGGCUUAG&quot;</span>,</span>
<span id="cb25-4"><a href="#cb25-4" tabindex="-1"></a>                <span class="st">&quot;CGUACGGUAGGCUUAACUGCGUACGAUCGAUCGC&quot;</span>)</span>
<span id="cb25-5"><a href="#cb25-5" tabindex="-1"></a></span>
<span id="cb25-6"><a href="#cb25-6" tabindex="-1"></a>GenePack<span class="sc">::</span><span class="fu">setProductSequence</span>(lncrna_gene) <span class="ot">&lt;-</span> seq</span>
<span id="cb25-7"><a href="#cb25-7" tabindex="-1"></a>GenePack<span class="sc">::</span><span class="fu">getProductSequence</span>(lncrna_gene)</span></code></pre></div>
<pre><code>## AUGCUUAGCGUACGGUAGGCUUAACUGCGUACGAUCGAUCGGCUAAUCGGCUUAGCGUACGGUAGGCUUAACUGCGUACG 
## AUCGAUCGGCUAAGCGUACGGUAGGCUUAACUGCGUACGAUCGAUCGGCUUAGCGUACGUAGGCUCAACUGCGUACGAUC 
## GAUCGGCUUAGCGUACGGUAGGCUUAACUGCGGACGAUCGAUCGGCUUAGCGUACGGUAGGCUUAACUGCGUACGAUCGA 
## UCGC</code></pre>
</div>
</div>
<div id="computing-the-product-gene-length" class="section level1">
<h1>Computing the product gene length</h1>
<p>An example regarding the use of the <strong>lengthProduct</strong>
function is provided below. The function is applied on the lncrna_gene
object previously created. In the same way, the function can be applied
to all the other gene objects. It computes the gene product sequence
length counting the number of nucleotides or amino acids, depending on
the gene object.</p>
<div class="sourceCode" id="cb27"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" tabindex="-1"></a>computed_length <span class="ot">&lt;-</span> GenePack<span class="sc">::</span><span class="fu">lengthProduct</span>(lncrna_gene)</span></code></pre></div>
</div>
<div id="displaying-the-gene-object-information" class="section level1">
<h1>Displaying the gene object information</h1>
<p>An example regarding the use of the <strong>showGeneObject</strong>
function is shown. The function is applied on the lncrna_gene object
previously created. In the same way, the function can be applied to all
the other gene objects.</p>
<div class="sourceCode" id="cb28"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" tabindex="-1"></a>GenePack<span class="sc">::</span><span class="fu">showGeneObject</span>(lncrna_gene)</span></code></pre></div>
<pre><code>## Gene type: LncRNAGene
## Ensembl ID: ENST00005859750
## HUGO Symbol: NEWSYMBOL1
## Name: lncRNA gene name
## Description: gene description
## Tissue Specificity: liver, small bowel
## Gene Structure:
## GRanges object with 1 range and 0 metadata columns:
##       seqnames    ranges strand
##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;
##   [1]     chr1  205-1210      +
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths
## Gene Product:
##   lncrna_id: 
##   newlncRNAID
##   lncrna_sequence: 
##   AUGCUUAGCGUACGGUAGGCUUAACUGCGUACGAUCGAUCGGCUAAUCGGCUUAGCGUACGGUAGGCUUAACUGCGUACG
##   AUCGAUCGGCUAAGCGUACGGUAGGCUUAACUGCGUACGAUCGAUCGGCUUAGCGUACGUAGGCUCAACUGCGUACGAUC
##   GAUCGGCUUAGCGUACGGUAGGCUUAACUGCGGACGAUCGAUCGGCUUAGCGUACGGUAGGCUUAACUGCGUACGAUCGA
##   UCGC
## Clinical Significance: association with bowel cancer
## -----------------</code></pre>
</div>
<div id="conclusion" class="section level1">
<h1>Conclusion</h1>
<p>The <strong>GenePack</strong> package provides a comprehensive set of
S4 classes to represent various human gene types, including both
protein-coding and non-coding genes, along with a series of functions to
create, manage and modify the gene objects created. This vignette has
provided users crucial information and examples to effectively use this
package. For more detailed information regarding the classes, functions,
and additional examples, please refer to the package documentation.</p>
</div>
<div id="session-info" class="section level1">
<h1>Session Info</h1>
<div class="sourceCode" id="cb30"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" tabindex="-1"></a><span class="fu">sessionInfo</span>()</span></code></pre></div>
<pre><code>## R version 4.3.1 (2023-06-16 ucrt)
## Platform: x86_64-w64-mingw32/x64 (64-bit)
## Running under: Windows 10 x64 (build 19045)
## 
## Matrix products: default
## 
## 
## locale:
## [1] LC_COLLATE=Italian_Italy.utf8  LC_CTYPE=Italian_Italy.utf8   
## [3] LC_MONETARY=Italian_Italy.utf8 LC_NUMERIC=C                  
## [5] LC_TIME=Italian_Italy.utf8    
## 
## time zone: Europe/Rome
## tzcode source: internal
## 
## attached base packages:
## [1] stats     graphics  grDevices utils     datasets  methods   base     
## 
## other attached packages:
## [1] GenePack_1.0.0
## 
## loaded via a namespace (and not attached):
##  [1] cli_3.6.1               knitr_1.47              rlang_1.1.2            
##  [4] xfun_0.44               jsonlite_1.8.8          S4Vectors_0.40.2       
##  [7] RCurl_1.98-1.14         htmltools_0.5.8.1       sass_0.4.9             
## [10] stats4_4.3.1            rmarkdown_2.27          evaluate_0.24.0        
## [13] jquerylib_0.1.4         bitops_1.0-7            fastmap_1.2.0          
## [16] yaml_2.3.7              IRanges_2.36.0          lifecycle_1.0.4        
## [19] GenomeInfoDb_1.38.8     compiler_4.3.1          XVector_0.42.0         
## [22] rstudioapi_0.16.0       digest_0.6.33           R6_2.5.1               
## [25] GenomeInfoDbData_1.2.11 GenomicRanges_1.54.1    bslib_0.7.0            
## [28] tools_4.3.1             zlibbioc_1.48.2         BiocGenerics_0.48.1    
## [31] cachem_1.1.0</code></pre>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
